### PR TITLE
Support a flag to disable md5 comparisons during backups & Azure multipart MD5 fix

### PR DESCRIPTION
**Pull request by mkcello96**
This allows medusa users to optionally disable the md5 checks that Medusa does during backups (the checks will remain enabled by default). 
To explain in further detail what this flag does: during node backups to cloud storage, if a file is found in the cache AND that file is found to have the same size and MD5 hash as the one that is found in the manifest, the local file is presumed to be same as the one in the manifest, and therefore can be skipped during backup.

When this flag is specified in a backup command, the only change to the above behavior is that the MD5 has check is skipped when deciding if the local file is the same as the one in the manifest. 
I expect some medusa users will be OK with removing these MD5 checks, as SSTables are immutable.

In my experience enabling this flag sped up differential backups by more than 10x.

In addition, there is an Azure-specific fix in this PR for the multipart MD5 check. Prior to this PR,[ Azure was using a different method of computing MD5s for multipart files](https://github.com/thelastpickle/cassandra-medusa/blob/master/medusa/storage/azure_storage.py#L98), which followed AWS standards. These are not followed in Azure. Further, [az-cli only uses the common way of computing MD5s when uploading blobs to the cloud](https://github.com/thelastpickle/cassandra-medusa/blob/master/medusa/storage/azure_blobs_storage/azcli.py#L71). Therefore there was a mismatch in MD5 hashes between those in the manifest and those calculated when a file was found in the cache, causing re-uploads of all multipart files during backups.

### CODE REVIEW

**Review by rzvoncek — DISMISSED** [review]
The changes look good to me. However, I'd still like a second pair of eyes. @adejanovski , could you please have a look once you have a moment?

**Review by adejanovski — COMMENTED** [review]
It's great to have that PR in @mkcello96, thanks!

I have a few thoughts to share here:

- I think it would be good to make this a config setting primarily, and possibly make it possible to override it with a flag
- the `verify` command will perform hash checks as well through the `blob_matches_manifest()` method, which should be subject to the same setting (if backups are taken without checking hashes, we can assume the check should be disabled at verify time, unless requested otherwise explicitly)
- I'm wondering if maybe we could make this the default behavior, with an option to enable md5 checks. I know that hash computation takes quite a lot of time at backup time for everyone and I can think of a single case where it could be useful after restoring an old backup. If md5 checks were disabled, we could then advise to run a full backup instead of a differential one after the restore.

@rzvoncek, wdyt?

**Comment by mkcello96** [comment]
@adejanovski I like all of those ideas. Just FYI, for our medusa usage we're relying on the `verify` command simply to validate MD5 hashes if needed. It's an audit/security requirement to be able to perform that check. 
But for backups there's no need for the MD5 checks. 
So our config would have MD5 checks disabled, but I can flip the boolean of the flag in this PR to be able to enable it during `verify`s (as you suggested).

**Comment by mkcello96** [comment]
@adejanovski I added the changes you discussed. Let me know what you think. Happy to make changes for the wiki pages as well if needed.
Thanks!

**Comment by mkcello96** [comment]
I tried to please the sonarcloud checker but I don't see a reasonable way to prevent the duplicate lines in the integration_steps file. I think it is mostly complaining that I added `"enable_md5_checks": "false"` 3 times for each of the configs defined in the tests.
@adejanovski Let me know if I should look into this more. From what I understand these sonarcloud checks are only for new code added in this PR, so the build failure here should not influence other PRs after merging.

File: tests/integration/features/integration_tests.feature
Comment (mkcello96) [inline tests/integration/features/integration_tests.feature:39]: Alternatively, I can add 2 different tests if that is more desirable. One for just `I can verify the backup named "first_backup" `, with md5 disabled, and another for `I can verify the backup named "first_backup" with md5 enabled successfully`

File: tests/integration/features/integration_tests.feature
Comment (adejanovski) [inline tests/integration/features/integration_tests.feature:614]:
```suggestion
        When I perform a backup in "differential" mode of the node named "first_backup" with md5 checks "disabled"
        Then I can verify the backup named "first_backup" with md5 checks "disabled" successfully
```

File: tests/integration/features/integration_tests.feature
Comment (adejanovski) [inline tests/integration/features/integration_tests.feature:660]:
```suggestion
        When I perform a backup in "full" mode of the node named "first_backup" with md5 checks "disabled"
        When I perform a backup in "differential" mode of the node named "second_backup" with md5 checks "disabled"
        Then I can verify the backup named "first_backup" with md5 checks "disabled" successfully
        Then I can verify the backup named "second_backup" with md5 checks "disabled" successfully
```

File: tests/integration/features/steps/integration_steps.py
Comment (adejanovski) [inline tests/integration/features/steps/integration_steps.py:603]: In order to avoid duplications (which indeed make Sonar unhappy), we could define a global variable at the top:
```
config_checks = {"health_check": "cql", "enable_md5_checks": "false"}
```

And then use: 

```
     config["checks"] = config_checks
```

On these lines and also on line 731.

File: tests/integration/features/steps/integration_steps.py
Comment (adejanovski) [inline tests/integration/features/steps/integration_steps.py:831]:
```suggestion
@when(r'I perform a backup in "{backup_mode}" mode of the node named'
      r' "{backup_name}" with md5 checks "{md5_enabled}"')
def _i_perform_a_backup_of_the_node_named_backupname(context, backup_mode, backup_name, md5_enabled):
```

File: tests/integration/features/steps/integration_steps.py
Comment (adejanovski) [inline tests/integration/features/steps/integration_steps.py:833]:
```suggestion
        = medusa.backup_node.main(context.medusa_config, backup_name, None, 
                                                            True if md5_enabled == "enabled" else False, backup_mode)
```

File: tests/integration/features/steps/integration_steps.py
Comment (adejanovski) [inline tests/integration/features/steps/integration_steps.py:997]:
```suggestion
@then(r'I can verify the backup named "{backup_name}" with md5 checks "{md5_enabled}" successfully')
def _i_can_verify_the_backup_named_successfully(context, backup_name, md5_enabled):
    medusa.verify.verify(context.medusa_config, backup_name, True if md5_enabled == "enabled" else False)
```

**Review by adejanovski — CHANGES_REQUESTED** [review]
Thanks a lot for handling this work @mkcello96, I've been trying to find someone to add this flag for a while now 😅 

I've suggested a few minor changes to the Gherkin wording to make it more natural to read and also a few recommendations on using Python ternaries rather than classic if/else structures.
Otherwise it looks great to me!

Would you, by any chance, have some numbers on the time savings that can be expected when disabling the hash checks?

File: tests/integration/features/steps/integration_steps.py
Comment (mkcello96) [reply tests/integration/features/steps/integration_steps.py]: Thanks that worked :)

**Comment by mkcello96** [comment]
> > @adejanovski I added the changes you discussed. Let me know what you think. Happy to make changes for the wiki pages as well if needed.
> > Thanks!
> 
> Thanks for offering to update the wiki, we're currently in the process of moving the docs back in tree in #317 (using a wiki was my mistake in the first place, it does not allow pushing new features and the associated documentation in the same PR, which is counterproductive). I think we should be able to merge that PR today, which will allow you to rebase on top of `master` and update the docs right in this PR.

Thanks for the review! I'll keep an eye on that PR and rebase after that's in.

**Comment by mkcello96** [comment]
@adejanovski Readme changes are in!

**Review by adejanovski — APPROVED** [review]
Awesome work @mkcello96, thanks a lot!

**Pull request merged**
